src/evict/evict_inline.h

```diff
@@ -101,7 +127,72 @@ __wt_evict_page_soon(WT_SESSION_IMPL *session, WT_REF *ref)
 {
     WT_UNUSED(session);
 
-    __wt_atomic_store64(&ref->page->read_gen, WT_READGEN_OLDEST);
+    __wt_atomic_store64(&ref->page->read_gen, WT_READGEN_EVICT_SOON);
+}
+
+/*
+ * __wt_evict_page_first_dirty --
+ *     Tell eviction if this is the first time the page has been modified while in the cache. The
+ *     eviction mechanism will then update the page's eviction state as needed. In this function,
+ *     although the page was not initially required, it has now been modified, so we prefer to
+ *     retain it instead of evicting it immediately.
+ */
+static WT_INLINE void
+__wt_evict_page_first_dirty(WT_SESSION_IMPL *session, WT_PAGE *page)
+{
+    /*
+     * In the event we dirty a page which is flagged for eviction soon, we update its read
+     * generation to avoid evicting a dirty page prematurely.
+     */
+    if (__wt_atomic_load64(&page->read_gen) == WT_READGEN_WONT_NEED)
+        __wti_evict_read_gen_new(session, page);
+}
+
+/*
+ * __wt_evict_touch_page --
+ *     Tell eviction when we touch a page so it can update its eviction state for that page. The
+ *     caller may set flags indicating that it doesn't expect to need the page again or that it's an
+ *     internal read. The latter is used by operations such as compact, and eviction, itself, so
+ *     internal operations don't update page's eviction state.
+ */
+static WT_INLINE void
+__wt_evict_touch_page(WT_SESSION_IMPL *session, WT_PAGE *page, bool internal_only, bool wont_need)
+{
+    /* Is this the first use of the page? */
+    if (__wt_atomic_load64(&page->read_gen) == WT_READGEN_NOTSET) {
+        if (wont_need)
+            __wt_atomic_store64(&page->read_gen, WT_READGEN_WONT_NEED);
+        else
+            __wti_evict_read_gen_new(session, page);
+    } else if (!internal_only)
+        __wti_evict_read_gen_bump(session, page);
+}
+
+/*
+ * __wt_evict_page_init --
+ *     Initialize page's eviction state for a newly created page.
+ */
+static WT_INLINE void
+__wt_evict_page_init(WT_PAGE *page)
+{
+    __wt_atomic_store64(&page->read_gen, WT_READGEN_NOTSET);
+}
+
+/*
+ * __wt_evict_inherit_page_state --
+ *     When creating a new page from an existing page, for example during split, initialize the read
+ *     generation on the new page using the read generation of the original page, unless this was a
+ *     forced eviction, in which case we leave the new page with the default initialization.
```

Comment:
```suggestion
 *     immediate eviction, in which case we leave the new page with the default initialization.
```
nit: WONT_NEED won't be put in the urgent queue so "forced" seems like too strong a word here

Reply:
Explained above -
forced eviction -> (oldest | wont_need)